[4.0] Fixes radio fields layout

YatharthVyas opened the pull request:
Pull Request for Issue #33144

### Summary of Changes
This PR changes the CSS class and styling applied to the input type=radio's label.  Refer to https://github.com/joomla/joomla-cms/blob/2c302641aaad823ac91421a0b11f14e3f8e2fd27/layouts/joomla/form/field/radio/buttons.php#L49
Bootstrap suggests that you organize the radio options as:
```
<div class="form-check">
  <input class="form-check-input" type="radio" name="exampleRadios" id="exampleRadios1" value="option1" checked>
  <label class="form-check-label" for="exampleRadios1">
    Default radio
  </label>
</div>
<div class="form-check">
  <input class="form-check-input" type="radio" name="exampleRadios" id="exampleRadios2" value="option2">
  <label class="form-check-label" for="exampleRadios2">
    Second default radio
  </label>
</div>
```
This PR changes the `form-check` class applied to each option's `<label>`  to `form-check-label` and encloses each option's input-label pair in a `<div class="form-check">`


### Testing Instructions
Change the XML form for any plugin's admin console page to include a radio field like: (similar to the [example in Joomla 4 docs](https://docs.joomla.org/J4.x:Creating_a_Plugin_for_Joomla))
```
<field
  name="search_archived"
  type="radio"
  class="switcher"
  label="LABEL_NAME"
  default="0"
>
  <option value="1">JYES</option>
  <option value="0">JNO</option>
</field>
```
Note that there is **NO** `layout="joomla.form.field.radio.switcher"` part.
Please also test for `class="btn-group"` and `class="btn-group-yesno"`

For example, in the screenshots below I've removed line 37 from plugins/content/vote/vote.xml to achieve the desired testing condition
https://github.com/joomla/joomla-cms/blob/2c302641aaad823ac91421a0b11f14e3f8e2fd27/plugins/content/vote/vote.xml#L37

### Actual result BEFORE applying this Pull Request
![image](https://user-images.githubusercontent.com/53610833/115119355-8ee9e900-9fc5-11eb-8f9b-81394bc4bef8.png)



### Expected result AFTER applying this Pull Request

### `no class`
![image](https://user-images.githubusercontent.com/53610833/115140139-6f010680-a053-11eb-8264-86fef2589e6f.png)

### `btn-group`
![image](https://user-images.githubusercontent.com/53610833/115139959-7e338480-a052-11eb-878f-bd153ac5f321.png)

### `btn-group-yesno`
![image](https://user-images.githubusercontent.com/53610833/115139972-8c81a080-a052-11eb-8233-dec951fbccd9.png)


### Documentation Changes Required
 `layout="joomla.form.field.radio.switcher"` should be mentioned in the [Joomla 4 Plugin Tutorial Page's XML Section](https://docs.joomla.org/J4.x:Creating_a_Plugin_for_Joomla) for making radio fields with 2 options.

YatharthVyas committed:
Removed extra styling applied to label

Removed extra styling applied to `<label>` and enclosed each option in a `<div>`

rjharishabh commented:
I have tested this item ✅ successfully on 69ce6cb

particthistle commented:
I have tested this item :white_check_mark: successfully on 69ce6cbb82e5a0e949c97d409d1cd8d39df21053<br /><br />Works as described.

Nice work @YatharthVyas - is that your first PR that's going to be merged for Joomla? If so, welcome on board and I look forward to seeing more soon.

Documentation Required label should be added: The layout parameter should be included in the documentation if it's not there as that should become best practice for extension developers going forward to give a consistent UX in the backend.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

particthistle commented:
RTC.

@richard67 not sure what's happened there with the first test by @rjharishabh not triggering JTracker properly?

rjharishabh commented:
> RTC.
> 
> @richard67 not sure what's happened there with the first test by @rjharishabh not triggering JTracker properly?

@particthistle commenting from github

rjharishabh commented:
I have tested this item ✅ successfully on 69ce6cb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

rjharishabh commented:
I have tested this item :white_check_mark: successfully on 69ce6cbb82e5a0e949c97d409d1cd8d39df21053<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

YatharthVyas commented:
@particthistle @rjharishabh, Thanks for testing!

> Nice work @YatharthVyas - is that your first PR that's going to be merged for Joomla? If so, welcome on board and I look forward to seeing more soon.

It's my second one that's about to be merged. I'm glad that you found it helpful and thanks for the kind words! I hope to contribute more

Fedik commented:
Sorry, the fix is incorrect.

> This PR removes the extra styling applied to each option's `<label>`

You have removed possibility to set a class per `option`, that may come from `$option->class`

Need to check what exactly cause the issue, which class or which style, to make a proper fix.

YatharthVyas commented:
> Sorry, the fix is incorrect.
> 
> > This PR removes the extra styling applied to each option's `<label>`
> 
> You have removed possibility to set a class per `option`, that may come from `$option->class`
> 
> Need to check what exactly cause the issue, which class or which style, to make a proper fix.
> 
> ~I can give you a tip: Look in to `if ($isBtnYesNo)` section, it not used in Joomla 4, the issue may come from there.~
> My tip is wrong sorry :)

The class is still being applied to the checkbox, it's just the label that is not re-using it again. I believe if you want to allow the users to add classes to both the option and label separately then there should be 2 different parameters for them. Doing so would require changes in the XML schema. 

Also, the Joomla 3 docs ([Link](https://docs.joomla.org/Radio_form_field_type)) say that the class is only used to make 'nice coloured buttons' so it feels like the style must only be applied on the buttons, no?

Fedik commented:
> The class is still being applied to the checkbox

That is correct, if `$option->class` exists then it should be applied.

The issue comes from class `.form-check`, which applied to label, that is incorrect.
Bootstrap doc: https://getbootstrap.com/docs/5.0/forms/checks-radios/#default-stacked

That looks like the markup of radiobutons may need to review, to fit the doc.
You can try to look in to the checboxes layout `joomla.form.field.checkboxes`, there seems correct markup

Or maybe someone from frontender can give you a tip.
@ciar4n do you have a tip here? ;)

brianteeman commented:
There is something not right with this PR. You have removed $optionClass from the label but that would appear to have been the only place it was used so you should also have removed all the code that created it as well.

YatharthVyas committed:
makes options bootstrap friendly and reverts $optionClass

YatharthVyas commented:
## Summary of modifications:
Added the `class="<?php echo trim($optionClass . ' ' . $style); ?>"` back to the `label`
And made the layout bootstrap friendly
thanks @Fedik for the help! 

## Screenshots after the fix:
### btn-group
![image](https://user-images.githubusercontent.com/53610833/115139972-8c81a080-a052-11eb-8233-dec951fbccd9.png)
### btn-yes-no
![image](https://user-images.githubusercontent.com/53610833/115139959-7e338480-a052-11eb-878f-bd153ac5f321.png)
### Regular
![image](https://user-images.githubusercontent.com/53610833/115140139-6f010680-a053-11eb-8264-86fef2589e6f.png)
### layout="joomla.form.field.radio.switcher"
![image](https://user-images.githubusercontent.com/53610833/115140245-fa7a9780-a053-11eb-8ff5-a7a94dc5f71a.png)

YatharthVyas committed:
fix indentation

particthistle commented:
I have tested this item :white_check_mark: successfully on 7b6486b66b87bbf7859cbd13d5e00bb8343c3b9e<br /><br />Tested again. Changes now do what the changes do, even more correctly than earlier in the day.

All four options tested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

rjharishabh commented:
I have tested this item :white_check_mark: successfully on 7b6486b66b87bbf7859cbd13d5e00bb8343c3b9e<br /><br />I have tested this item ✅ successfully on 7b6486b

All four options tested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

Quy commented:
The styling/markup for button group is not correct. See https://getbootstrap.com/docs/5.0/components/button-group/

richard67 commented:
@Quy So I shall remove "RTC" and add "Updates Requested"?

Quy commented:
Yes please.

richard67 commented:
Back to pending, see previous comments.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

richard67 commented:
@Quy What doesn't match in particular?

Quy commented:
No spacing between buttons.

YatharthVyas committed:
fix html for .btn-group

`<div class="form-check">` is conditionally rendered only if class is not `btn-group`
Class applied to radio `<input>` is conditionally decided between `btn-check` and `form-check`

YatharthVyas commented:
> No spacing between buttons.

Thanks, fixed.
```
<div class="btn-group radio">
   <input class="btn-check" type="radio" id="jform_params_show_total_votes0" name="jform[params][show_total_votes]" value="0" checked="checked">
   <label for="jform_params_show_total_votes0" class="btn btn-outline-secondary"> Hide </label>
   <input class="btn-check" type="radio" id="jform_params_show_total_votes1" name="jform[params][show_total_votes]" value="1" >
   <label for="jform_params_show_total_votes1" class="btn btn-outline-secondary">Show </label>
</div>
```

![image](https://user-images.githubusercontent.com/53610833/115147452-a7ffa200-a078-11eb-8b27-fcf49439c795.png)

YatharthVyas committed:
fix spacing for btn-group-yesno

Apply the `.btn-group` instead of `.btn-group-yesno` class to wrapper div if class="btn-group-yesno"

Quy commented on layouts/joomla/form/field/radio/buttons.php:
```diff
@@ -84,46 +87,48 @@
-			$style       = $disabled ? 'style="pointer-events: none"' : '';
+			<?php echo $blockStart; ?>
+				<?php
+				$disabled    = !empty($option->disable) ? 'disabled' : '';
+				$style       = $disabled ? 'style="pointer-events: none"' : '';
```
```suggestion
				$disabled = !empty($option->disable) ? 'disabled' : '';
				$style    = $disabled ? 'style="pointer-events: none"' : '';
```

Commit:
Update layouts/joomla/form/field/radio/buttons.php

particthistle commented:
I have tested this item :white_check_mark: successfully on f58aaf33bce1f25fed2ac4da09d2e0ed97f858ed<br /><br />I did think the button groups didn't quite look right last test... that explains why now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

chmst commented:
I have tested this item :white_check_mark: successfully on f58aaf33bce1f25fed2ac4da09d2e0ed97f858ed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33176">issues.joomla.org/tracker/joomla-cms/33176</a>.</sub>

YatharthVyas commented:
Thank you for merging. Could you please add the Documentation Changes label for 

>`layout="joomla.form.field.radio.switcher"` should be mentioned in the [Joomla 4 Plugin Tutorial Page's XML Section](https://docs.joomla.org/J4.x:Creating_a_Plugin_for_Joomla) for making radio fields with 2 options.

richard67 commented:
@YatharthVyas But in your description in section "Documentation Changes Required" you wrote:

> None, if the PR is merged.

YatharthVyas commented:
Apologies but it was later that I realized that class="switcher" (the current class attribute mentioned in Joomla 4 Plugin Docs XML section) isn't a valid attribute. It should be one of the following :

- `layout="joomla.form.field.radio.switcher"`
- class="btn-group"
- class="btn-group-yesno"

I forgot the edit and correct this part.

regularlabs commented:
This PR removes the ability to set custom class names.
It is now not able to do something like:
```
<field name="..." type="radio"
        class="btn-group btn-group-yesno my-own-class">
```
